### ISSUE TITLE
No error when generic incompletely specified in subclass

### ISSUE THREAD

**Issue by joelberkeley-pio** [python/mypy#9036]
I'm reporting a bug. This doesn't error:
```python3
from abc import ABC, abstractmethod
from typing import TypeVar, Generic

T = TypeVar('T')

class Base(ABC, Generic[T]):
    @abstractmethod
    def foo(self, o: T) -> None:
        pass

class Sub(Base[str]):
    def foo(self, o: T) -> None:
        return None

Sub().foo(1)
```
I'd have expected it to error on two counts:
  - `Sub.foo` should require `T` to be `str` since we've specified `Base[str]` as the parent
  - `Sub.foo(1)` should error saying argument should be `str` not `int` (curiously PyCharm does flag this up)

mypy 0.761

>  Do you see the same issue after installing mypy from Git master?

yes

mypy.ini
```
[mypy]
ignore_missing_imports = True
strict_optional = False
```

**Comment by ilevkivskyi** [comment]
This code is technically type-safe, since `def [T](T) -> None` is a subtype of `def (str) -> None`.

**Comment by joelberkeley-pio** [comment]
ok, I think I can see how that specific example type checks, but I certainly feel it's a concern.

For example, it leaves it open to mistakes. If I wanted to change the implementation later in a way that assumes `o` is a `str`, I'll need to update the API when the type checker points out that `print("some text: " + o)` can't add `T` and `str`. Naturally, this can break users' code.

Related, the following Scala code doesn't compile
```
trait Base[T] {
  def foo(o: T): Unit
}

class Sub extends Base[String] {
  def foo[T](o: T): Unit = {}
}
```
with error
> error: class Sub needs to be abstract, since method foo in trait Base of type ( o: String)scala.this.Unit is not defined
